src/Socket_p.h

```diff
@@ -284,203 +243,192 @@ namespace Arcus
-            if(ret == -1)
+            result = platform_socket.readBytes(current_message->getRemainingSize(), &current_message->data[current_message->received_size]);
+
+            if(result == -1)
             {
             #ifndef _WIN32
                 if(errno != EAGAIN)
             #endif
                 {
-                    delete[] message.data;
-                    message.data = nullptr;
-                    message.size_recv = 0;
-                    message.state = MESSAGE_STATE_INIT;
+                    current_message.reset();
                 }
             }
             else
             {
-                message.size_recv += ret;
-                if(message.size_recv >= message.size)
+                current_message->received_size = current_message->received_size + result;
+                if(current_message->isComplete())
                 {
-                    if (message.valid) {
-                        message.state = MESSAGE_STATE_DISPATCH;
-                    }
-                    else
+                    if(!current_message->valid)
                     {
-                        delete[] message.data;
-                        message.data = nullptr;
-                        message.state = MESSAGE_STATE_INIT;
+                        current_message.reset();
+                        return;
                     }
+
+                    current_message->state = WireMessage::MessageState::Dispatch;
                 }
             }
         }
 
-        if (message.state == MESSAGE_STATE_DISPATCH)
-        {
-            handleMessage(message.type, message.size, message.data);
-            delete[] message.data;
-            message.data = nullptr;
-            message.state = MESSAGE_STATE_INIT;
-        }
-    }
-
-    int SocketPrivate::readInt32(int32_t *dest)
-    {
-        int32_t buffer;
-        int num = ::recv(socketId, reinterpret_cast<char*>(&buffer), 4, 0);
-
-        if(num != 4)
-        {
-            return -1;
-        }
-
-        *dest = ntohl(buffer);
-        return 0;
-    }
-
-    int SocketPrivate::readBytes(int size, char* dest)
-    {
-        int num = ::recv(socketId, dest, size, 0);
-        if(num == -1)
-        {
-            return -1;
-        }
-        else
+        if (current_message->state == WireMessage::MessageState::Dispatch)
         {
-            return num;
+            handleMessage(current_message);
+            current_message.reset();
         }
     }
 
-    void SocketPrivate::handleMessage(int type, int size, char* buffer)
+    // Parse and process a message received on the socket.
+    void Socket::Private::handleMessage(const std::shared_ptr<WireMessage>& wire_message)
     {
-        if(messageTypes.find(type) == messageTypes.end())
+        if(!message_types.hasType(wire_message->type))
         {
-            errorString = "Unknown message type";
+            error(ErrorCode::UnknownMessageTypeError, "Unknown message type");
```

Comment:
We do not know the type. The only thing we could print is the type ID, which has become an implementation detail in the new API, so it does not necessarily make sense to expose that here.